pyro/ops/indexing.py

```diff
@@ -8,6 +8,76 @@ def _is_batched(arg):
     return isinstance(arg, torch.Tensor) and arg.dim()
 
 
+def _flatten(args, out):
+    if isinstance(args, tuple):
+        for arg in args:
+            _flatten(arg, out)
+    else:
+        # Combine consecutive Ellipsis.
+        if args is Ellipsis and out and out[-1] is Ellipsis:
+            return
+        out.append(args)
+
+
+def index(tensor, args):
+    """
+    Indexing with nested tuples.
+
+    See also the convenience wrapper :class:`Index`.
+
+    This is useful for writing indexing code that is compatible with multiple
+    interpretations, e.g. scalar evaluation, vectorized evaluation, or
+    reshaping.
+
+    For example suppose ``x`` is a parameter with ``x.dim() == 2`` and we wish
+    to generalize the expression ``x[..., t]`` where ``t`` can be any of:
+
+    - a scalar ``t=1`` as in ``x[..., 1]``;
+    - a slice ``t=slice(None)`` equivalent to ``x[..., :]``; or
+    - a reshaping operation ``t=(Ellipsis, None)`` equivalent to
+      ``x.unsqueeze(-1)``.
```

Comment:
I think that the code does its purpose. For reshaping, I think `(slice(None), None)` also works (that way, we don't need to combine Ellipses - which is aligned with the fact that at most 1 Ellipsis is used for indexing)

Reply:
The reason ``CompartmentalModel._vectorized_model()`` uses `t=(Ellipsis, None)` is to handle possible other dimensions, e.g. for multidimensional ``data`` tensors provided by users. It seemed easiest to me to use `.unsqueeze(-1)` semantics (via Ellipsis) rather than `.unsqueeze(1)` semantics (via colon): nonnegative indexing always gets me in trouble.

Reply:
Thanks for explaining!